In [7]:
import numpy as np
import pandas as pd
import Bags_Of_Words as bw
from sklearn.svm import LinearSVC
from sklearn import metrics
from sklearn.cross_validation import StratifiedKFold
from sklearn.feature_selection import RFECV
from sklearn.feature_selection import SelectKBest, chi2

In [17]:
train_fp = "../data/train_set10000.csv"
test_fp = "../data/test_set10000.csv"

BOW = bw.Bags_Of_Words(train_fp, test_fp, True)
print BOW['test'].shape
print BOW['train'].shape

(20000, 45661)
(80000, 45661)


In [18]:
df = pd.read_csv(train_fp)

In [13]:
# Create the RFE object and compute a cross-validated score.
svc = LinearSVC()
# The "accuracy" scoring is proportional to the number of correct
# classifications
rfecv = RFECV(estimator=svc, step=1, cv=StratifiedKFold(df.subreddit_index, 2),
              scoring='accuracy')
rfecv.fit(BOW['train'], df.subreddit_index)

print("Optimal number of features : %d" % rfecv.n_features_)

Optimal number of features : 21120


In [15]:
ch2 = SelectKBest(chi2, k=21120)
train = ch2.fit_transform(BOW['train'], df.subreddit_index)
test = ch2.transform(BOW['test'])

clf = LinearSVC().fit(train, df.subreddit_index)
predicted = clf.predict(test)

In [16]:
df_test = pd.read_csv(test_fp)

print np.mean(predicted == df_test.subreddit_index)

0.6606


## Results

- 1,000  - 71.0%  Optimal # of Features: 7520
- 5,000  - 66.06% Optimal # of Features: 21120
- 10,000 - 
- 20,000 - 
- 50,000 - 

## Results - Default 
Default parameters => penalty='l2', loss='squared_hinge', dual=True, tol=0.0001, C=1.0, multi_class='ovr', fit_intercept=True, intercept_scaling=1, class_weight=None, verbose=0, random_state=None, max_iter=1000
- 1,000  - 72.3%
- 5,000  - 66.28%
- 10,000 - 64.005%
- 20,000 - 64.4925%
- 50,000 - 62.307%